stew/io2.nim

```diff
@@ -1312,3 +1361,112 @@ proc readAllChars*(pathName: string): IoResult[string] =
 proc readAllFile*(pathName: string): IoResult[seq[byte]] =
   ## Alias for ``readAllBytes()``.
   readAllBytes(pathName)
+
+proc lockFile*(handle: IoHandle, kind: LockType, offset: int,
+               size: int): IoResult[void] =
+  ## Apply shared or exclusive file segment lock for file handle ``handle`` and
+  ## range specified by ``offset`` and ``size`` parameters.
+  ##
+  ## ``kind`` - type of lock (shared or exclusive). Please not that only
+  ## exclusive locks has cross-platform compatible behavior. But exclusive
+  ## locks require ``handle`` to be opened for writing.
+  ##
+  ## ``offset`` - starting byte offset in the file where the lock should
+  ## begin.
+  ##
+  ## ``size`` - length of the byte range to be locked.
+  when defined(posix):
+    let ltype =
+      case kind
+      of LockType.Shared:
+        cshort(posix.F_RDLCK)
+      of LockType.Exclusive:
+        cshort(posix.F_WRLCK)
+    var flockObj = FlockStruct(ltype: ltype, lwhence: cshort(posix.SEEK_SET),
+                               start: offset, length: size)
+    while true:
+      let res = posix.fcntl(cint(handle), posix.F_SETLK, addr flockObj)
+      if res == -1:
+        let errCode = ioLastError()
+        if errCode == EINTR:
+          continue
+        else:
+          return err(errCode)
+      else:
+        return ok()
+  elif defined(windows):
+    let (lowOffsetPart, highOffsetPart) = makeUint32(uint64(offset))
+    let (lowSizePart, highSizePart) = makeUint32(uint64(size))
```

Comment:
These could trigger defect on Nim 1.6 when a negative offset / size is passed. 
On Nim 1.2 it would convert the bit pattern, but Windows would then interpret it as a large positive number.

Return early error if these are negative, or use `uint64` in the args (to match `getFileSize` return type)